File: clients/web/src/app/shared/utils/workflow-sla.util.ts
export type WorkflowSlaLevel = 'ok' | 'warning' | 'overdue' | 'none';

/** SLA badge: green (>24h), amber (<24h), red (overdue). */
export function workflowSlaLevel(dueAt: string | null | undefined, nowMs = Date.now()): WorkflowSlaLevel {
  if (!dueAt) {
    return 'none';
  }
  const due = Date.parse(dueAt);
  if (Number.isNaN(due)) {
    return 'none';
  }
  const hoursLeft = (due - nowMs) / (1000 * 60 * 60);
  if (hoursLeft < 0) {
    return 'overdue';
  }
  if (hoursLeft < 24) {
    return 'warning';
  }
  return 'ok';
}

File: clients/mobile/lib/utils/workflow_sla_util.dart
enum WorkflowSlaLevel { ok, warning, overdue, none }

WorkflowSlaLevel workflowSlaLevel(String? dueAt, [DateTime? now]) {
  if (dueAt == null || dueAt.isEmpty) {
    return WorkflowSlaLevel.none;
  }
  final due = DateTime.tryParse(dueAt);
  if (due == null) {
    return WorkflowSlaLevel.none;
  }
  final reference = now ?? DateTime.now().toUtc();
  final hoursLeft = due.difference(reference).inMinutes / 60.0;
  if (hoursLeft < 0) {
    return WorkflowSlaLevel.overdue;
  }
  if (hoursLeft < 24) {
    return WorkflowSlaLevel.warning;
  }
  return WorkflowSlaLevel.ok;
}
